regnety/dataset/imagenet.py

```diff
@@ -0,0 +1,223 @@
+        "synset": tf.io.FixedLenFeature([], tf.string),
+}
+    
+
+class ImageNet:
+    """Class for all ImageNet related functions, includes TFRecords loading,
+    preprocessing and augmentations. TFRecords must follow the format given
+    below. If not specified otherwise in `augment_fn` argument, following 
+    augmentations are applied to the dataset:
+    - Random sized crop (train only)
+    - Scale and center crop
+
+    Args:
+        tfrecs_filepath: list of filepaths of all TFRecords files
+        batch_size: batch_size for the Dataset
+        image_size: final image size of the images in the dataset
+        augment_fn: function to apply to dataset after loading raw TFrecords
+        num_classes: number of classes
+        randaugment: True if RandAugment is to be applied after other 
+            preprocessing functions. It is applied even if this is True and 
+            augment_fn is not 'default'. It is not applied in any case if this
+            argument is False.
+    """
+    def __init__(
+        self,
+        tfrecs_filepath: List[str]  = None,
+        batch_size: int = 128,
+        image_size: int = 224,
+        augment_fn: Union[str, Callable]  = "default",
+        num_classes: int = 10,
+        randaugment: bool = True,
+    ):
+
+        if (tfrecs_filepath is None) or  (tfrecs_filepath == []):
+            raise ValueError("List of TFrecords paths cannot be None or empty")
+        self.tfrecs_filepath = tfrecs_filepath
+        self.batch_size = batch_size
+        self.image_size = image_size
+        self.augment_fn = augment_fn
+        self.num_classes = num_classes
+        self.randaugment = randaugment
+        if self.randaugment:
+            self._augmenter = RandAugment(magnitude=5, num_layers=2)
+
+    def decode_example(self, example: tf.Tensor) -> dict:
+        """Decodes an example to its individual attributes
+
+        Args:
+            example: A TFRecord dataset example.
+
+        Returns:
+            Dict containing attributes from a single example. Follows
+            the same names as TFRECORDS_FORMAT.
+        """
+        image = tf.cast(tf.io.decode_jpeg(example["image"]), tf.float32)
+        height = example['height']
+        width = example['width']
+        filename = example["filename"]
+        label = example["label"]
+        synset = example["synset"]
+        return {
+            "image": image,
+            "height": height,
+            "width": width,
+            "filename": filename,
+            "label": label,
+            "synset": synset,
+        }
+
+    def _read_tfrecs(self) -> Type[tf.data.Dataset]:
+        """Function for reading and loading TFRecords into a tf.data.Dataset.
+
+        Returns:
+            A tf.data.Dataset
+        """
+        ds = tf.data.TFRecordDataset(self.tfrecs_filepath)
+        ds = ds.map(
+            lambda example: tf.io.parse_example(example, _TFRECS_FORMAT)
+        )
```

Comment:
Parallelize this `map()` call by specifying `num_parallel_calls=tf.data.AUTOTUNE`. [Here's](https://www.tensorflow.org/guide/data_performance) the official guide on `tf.data` to get the most out of it in terms of performance.

Reply:
Okay

Reply:
@sayakpaul 
AFAIK the map function records all the calls and employs a model to optimize it. This may be a silly question, but do we parallelize every function or some selective ones?

Reply:
> AFAIK the map function records all the calls and employs a model to optimize it. 

Could you expand a bit more on this with some credible sources? 

> This may be a silly question, but do we parallelize every function or some selective ones?

Not silly at all. If we mindlessly parallelize things then it may yield a negative effect. That said, in this case parallelizing the record parsing call won't be harmful.

Reply:
Sorry for delay in the reply, I didn't see the comment earlier.

>Could you expand a bit more on this with some credible sources?

"_The auto-tuning implementation uses the processing time
and the input pipeline structure to build an analytical model
that is used to estimate how input pipeline parameters affect
end-to-end latency_" - Read this from [here](https://arxiv.org/pdf/2101.12127.pdf).

> If we mindlessly parallelize things then it may yield a negative effect.

Could you please check the code once again? I have added `num_parallel_calls` almost everywhere. Also, I just realized,  is `random_sized_crop`  and `scale_and_center`  traceable functions? I mean, `example` is a dict which is a part of the tf.data.dataset but its individual image tensors may vary in size. So will the function execute without making too many graphs (and possibly OOM)?

Reply:
But we don't actually ask it to auto-tune it, it's won't, right? That is the entire point of using `AUTOTUNE`. 

> Could you please check the code once again? I have added num_parallel_calls almost everywhere. Also, I just realized, is random_sized_crop and scale_and_center traceable functions? I mean, example is a dict which is a part of the tf.data.dataset but its individual image tensors may vary in size. So will the function execute without making too many graphs (and possibly OOM)?

To confirm this you can debug the input bottleneck inside TensorBoard. Here's the guide: https://www.tensorflow.org/guide/data_performance.